### ISSUE TITLE
Try-It OAUTH2 Client Credentials XHR aborted

### ISSUE THREAD

**Issue by CameronGo** [mulesoft/api-console#292]
When I attempt to use Try-It with an OAUTH2 Client Credentials grant type the console immediately returns a response:
XHR aborted

The API endpoint does not receive any communications, but the oauth2 implementation appears to be returning a 405 response.

Does anyone have any idea what might be happening here?

**Comment by CameronGo** [comment]
Hopefully it helps to see what the oauth2 server is receiving, since I'm guessing that is where the problem is.  I setup some logging to output the headers and the POST data to the apache error log.  Below is the JSON representation of that output.  I don't see the client id and secret in the POST data, which is what the oauth2 server is expecting.  Is the api-console sending it some other way?

```json
[
    {
        "headers":[
            {
                "Host":"oauth2-dev.domain.com",
                "X-Real-IP":"321.321.321.321",
                "X-Forwarded-For":"123.123.123.123, 321.321.321.321",
                "X-Forwarded-Proto":"https",
                "Connection":"close",
                "Accept":"*\\/*",
                "Accept-Encoding":"gzip, deflate, sdch, br",
                "Accept-Language":"en-US,en;q=0.8",
                "Access-Control-Request-Headers":"accept, authorization, content-type",
                "Access-Control-Request-Method":"POST",
                "Origin":"https:\\/\\/chlorine:8446",
                "Referer":"https:\\/\\/chlorine:8446\\/api-console\\/dist\\/?raml=\\/pathtoraml\\/api.raml",
                "User-Agent":"Mozilla\\/5.0 (Windows NT 10.0; WOW64) AppleWebKit\\/537.36 (KHTML, like Gecko) Chrome\\/51.0.2704.84 Safari\\/537.36",
                "X-Forwarded-Port":"443"
            }
        ]
    },
    {
        "post":[
            [

            ]
        ]
    }
]
```


This compared to a working oauth2 request from POSTMAN:

```json
[
    {
        "headers":[
            {
                "Host":"oauth2-dev.domain.com",
                "X-Real-IP":"321.321.321",
                "X-Forwarded-For":"123.123.123, 321.321.321",
                "X-Forwarded-Proto":"https",
                "Connection":"close",
                "Content-Length":"160",
                "Accept":"*\\/*",
                "Accept-Encoding":"gzip, deflate, br",
                "Accept-Language":"en-US,en;q=0.8",
                "Cache-Control":"no-cache",
                "Content-Type":"application\\/x-www-form-urlencoded",
                "Origin":"chrome-extension:\\/\\/fhbjgbiflinjbdggehcddcbncdddomop",
                "Postman-Token":"596e0b74-ebeb-29f7-9a89-4c62769a7097",
                "User-Agent":"Mozilla\\/5.0 (Windows NT 10.0; WOW64) AppleWebKit\\/537.36 (KHTML, like Gecko) Chrome\\/51.0.2704.84 Safari\\/537.36",
                "X-Forwarded-Port":"443"
            }
        ]
    },
    {
        "post":[
            {
                "grant_type":"authorization_code",
                "client_id":"testclient",
                "client_secret":"testpass",
                "code":"a39ae713b4216f3d8fb19d7df2c40b368e043076",
                "redirect_uri":"http:\\/\\/www.google.com"
            }
        ]
    }
]
```

**Comment by CameronGo** [comment]
I've made some rough progress now, though not arrived at a solution yet.  It looks like the issue may be related to CORS on my oauth2 provider.  I setup CORS there and am now getting seeing more returned on my oauth2 provider, but it is not yet making it past that point.  In particular it looks like the client credentials auth from Try-It is attempting to pass in via headers rather than form data post, which is what I was doing with POSTMAN.  I'm still digging, but here are the headers I see on my oauth2 provider and the post data as it stands currently.

```json
[
    {
        "headers":[
            {
                "Host":"oauth2-dev.domain.com",
                "X-Real-IP":"321.321.321.321",
                "X-Forwarded-For":"123.123.123.123, 321.321.321.321",
                "X-Forwarded-Proto":"https",
                "Connection":"close",
                "Content-Length":"36",
                "Accept":"application\\/json, application\\/x-www-form-urlencoded",
                "Accept-Encoding":"gzip, deflate, br",
                "Accept-Language":"en-US,en;q=0.8",
                "Authorization":"Basic dGVzdGNsaWVudDp0ZXN0cGFzcw==",
                "Content-Type":"application\\/x-www-form-urlencoded",
                "Origin":"https:\\/\\/chlorine:8446",
                "Referer":"https:\\/\\/chlorine:8446\\/api-console\\/dist\\/?raml=\\/PursuantHealthv1\\/api.raml",
                "User-Agent":"Mozilla\\/5.0 (Windows NT 10.0; WOW64) AppleWebKit\\/537.36 (KHTML, like Gecko) Chrome\\/51.0.2704.84 Safari\\/537.36",
                "X-Forwarded-Port":"443"
            }
        ]
    },
    {
        "post":[
            {
                "scope":"",
                "grant_type":"client_credentials"
            }
        ]
    }
]
```

**Comment by CameronGo** [comment]
After a lot of tinkering and educating myself on CORS, the solution is working as expected.  For anyone else who comes along looking for similar input, I had to enable CORS and enable a rewite rule in my .htaccess file on both my OAUTH2 and API endpoint.  The issue I encountered after my enabling CORS was due to the fact I'd enabled the Access-Control-Allow-Origin header in both my apache2 conf file and my .htaccess file, which then returns two values for the Origin header and is not allowed.

For those who aren't familiar with CORS, as was my case, the Header always set ... affectively allows JavaScript from the api-console, that is running in your browser to query data from a URL host (Origin) that is different from that of the api-console.  In our case our api-portal is running on a host (dev.domain.com) so when that JS tries to communicate with oauth2.domain.com or api.domain.com those respective web servers do not allow the request.  Note: We did not implement the wildcard * for our Origin, but rather used a regex statement in the .htaccess that will effectively allow the use of multiple possible Origin values.

The latter issue with the RewriteRule is to address "Preflighted" requests.  There is a lot written up on these and how they work, so I won't get into that, but suffice to say the excerpt from the .htaccess file above worked to address this.

The last comment I'll add is regarding the use of a proxy.  I saw that mentioned in the documentation, but with no real context I couldn't determine how or why this would be used.  The answer is that if you implement both your api-console and your API under the same hostname (and in my case would have to include OAUTH2 provider) then they will all be from the same Origin.  In this case you do not need to enable CORS b/c the JavaScript requests from the one site will already be allowed to the others.  Our reverse proxy sitting in front of our APIs is running on nginx, so we could have deployed our api-console behind that as well, but have opted not to and so must have CORS enabled on the API.

Hopefully this gives a comprehensive answer to anyone looking into this in the future even if it does cover subjects that could be independently researched elsewhere.
